src/Swarm/Util/WindowedCounter.hs

```diff
@@ -0,0 +1,135 @@
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Sliding window for activity monitoring.
+module Swarm.Util.WindowedCounter (
+  WindowedCounter,
+
+  -- * Construction
+  mkWindow,
+
+  -- * Querying
+  getOccupancy,
+
+  -- * Maintenance
+  insert,
+  discardGarbage,
+) where
+
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Ratio
+import Data.Set (Set)
+import Data.Set qualified as Set
+import GHC.Generics (Generic)
+import Prelude hiding (length)
+
+-- | A "sliding window" of a designated span that supports insertion
+-- of tick "timestamps" that represent some state of interest during that tick.
+-- This data structure supports efficient querying of the ratio of
+-- {ticks for which that state existed}
+-- to the
+-- {total number of ticks spanned by the window}.
+--
+-- The primary use case is in displaying the "activity level" of a robot.
+--
+-- == Efficiency considerations
+--
+-- The data retention of the window shall be maintained externally by
+-- invoking the 'discardGarbage' function. However, we should not
+-- unconditionally invoke this function upon each game tick.
+--
+-- For efficiency, we do not want to iterate over every robot
+-- upon every tick; we only want to "visit" a robot if it is actually
+-- doing work that tick.
+-- Because of this, there may be some ticks in which the oldest element
+-- that is still stored falls outside of the nominal retention window.
+--
+-- One might think we could perform garbage collection whenever we execute queries.
+-- However, in the context in which the view powered by the query is generated, we
+-- are not permitted to mutate the "state" of the game
+-- (the type signature of the rendering function is @AppState -> Widget Name@).
+--
+-- Therefore, when we perform "queries" on the window, we must apply some
+-- filtering to exclude the "stragglers"; data members that have already fallen outside
+-- the window but have not yet been "garbage collected".
+-- We use a 'Set' to allow this filtering to be performed in @O(log n)@ time.
+--
+-- In the worst case, the entire dataset may "age out" without being garbage collected,
+-- so that an @O(log n)@ filtering operation might be performed upon every "frame refresh"
+-- of the UI view.
+-- However, we also store the largest element of the window separately from the 'Set' so that
+-- we can compare against it for a @O(1)@ short-circuited path once every member ages out.
+data WindowedCounter a = WindowedCounter
+  { _members :: Set a
+  , _lastLargest :: Maybe a
+  -- ^ NOTE: It is possible that '_lastLargest' may not exist in the 'Set'.
+  , _nominalSpan :: a
```

Comment:
It seems strange to me that the type parameter `a` is used both for members and for the nominal span.  That seems like it will end up being restricted to only being able to store sets of numbers.  And if that's so, I don't know what we gain by being polymorphic in `a`.  It might be more general to have truly arbitrary values of some type `a`, indexed by tick number, something like
```
  { _members :: Map TickNumber a
  , _lastLargest :: Maybe TickNumber
  , _nominalSpan :: Integer
  }
```

Reply:
What do you think of the latest type parameterization, @byorgey?

Reply:
It's better, though I guess I am still not sure what the polymorphism is buying us.  Do you have any concrete plans to instantiate `a` with anything other than a `TickNumber`?  As it stands, `a` still has to be `Offsettable` and `Ord` which pretty much limits it to be some kind of number.

Reply:
> Do you have any concrete plans to instantiate `a` with anything other than a `TickNumber`?

No concrete plans, but suppose it is also useable on Enums that aren't numbers.

In our codebase, I want to ensure that the API is used directly with `TickNumber` types to preserve the "safety" of the `TickNumber` newtype "end to end".  However, I also don't think that this `Swarm.Util.WindowedCounter` module should depend on the definition of `TickNumber` (which currently lives in the `Swarm.Game.CESK` module).

Therefore, I think it makes sense to keep `WindowedCounter` polymorphic in `a`.